### PR TITLE
Add `PageHeader` Component

### PR DESCRIPTION
**Pull request by jorytindall**
### :pushpin: Summary

If merged this PR adds the `PageHeader` component to Helios as a collaboration/contribution from the HCP design team and multiple stakeholders across product lines.

I added a variety of different use cases and edge cases to the showcase as a means to capture specific questions and "gotchas" with this component. Once approved, it probably makes sense to remove some of them, though I think there is value in including some of the contextual examples since this tows the line between a component and pattern. 

### 🔧  Details

There are several areas that I would like specific feedback on including:

* Potential issues with the Dropdown `listPosition` and lack of collision awareness
* Button order and alignment at smaller viewport sizes
* Responsive strategy
  * Using container queries; are able to use these yet, does the support exist?.
  * I've used `flex-wrap` with some of the inline elements, making the point at which they wrap variable. Is this something we want more control over?
  * Is there a more elegant way to handle the breakpoints without using static values? Considering the method in the side-nav to conditionally apply classes with JavaScript, is that a recommended method here? Should the classes be exposed to be overwritten?
  * The linter does not recognize container queries yet, is that an adjustment that needs to be made?
* I've written a few simple tests for this component, are more necessary? If so, I believe I can add them without too much trouble given that this is largely a presentational component.

### :link: External links

Jira ticket: [HDS-XXX](https://hashicorp.atlassian.net/browse/HDS-1892)
Figma file: [PageHeader](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/branch/Urm6MreKfPoJMMW1sWTR08/HDS-Product---Components?type=design&node-id=34664%3A70976&t=isQ1cpOW4GbGpAmh-1)

***

### 👀 Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (MelSumner) [inline packages/components/addon/components/hds/page-header/index.hbs:5]: is it possible to have the header be the outermost containing element?

We want to try to have all content contained in landmark elements; users with a screen reader can browse a page via landmarks, and a bookmark component outside of the header (one of the landmarks) is not ideal.

Comment (jorytindall) [reply packages/components/addon/components/hds/page-header/index.hbs:5]: The reason the header isn't the most outermost element is because it often wraps breadcrumbs which render a `nav` element. The intention was to not nest a landmark nav inside of a `header`.

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (jorytindall) [reply packages/components/addon/components/hds/page-header/index.hbs:11]: I believe this was the only way to account for the badge coupled with the title/heading, while also justifying the actions to the end of the container. It's a little bit div heavy though, I'll admit.

**Comment by KristinLBradley** [comment]
RE: Container queries

If we go by the [browser versions that Consul supports](https://github.com/hashicorp/consul/blob/main/ui/packages/consul-ui/config/targets.js), then it appears we can't use them. See: https://caniuse.com/css-container-queries

You could potentially ask if they would be willing to update their list of supported browsers to more recent ones though...

**Comment by MelSumner** [comment]
DOM order needs to match visual order; it's a known issue with flex-direction: reverse-row. I can add some extra links to this CRD so our team can read through. 👍

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (alex-ju) [reply packages/components/addon/components/hds/page-header/index.hbs:5]: This was my recommendation, Melanie, to avoid nesting them, but if you think nesting the two landmarks is not a concern we can revert and apply the header element at the top level of this component.

File: .changeset/light-seas-exercise.md
Comment (alex-ju) [inline .changeset/light-seas-exercise.md:5]:
```suggestion
Add `PageHeader` component
```

**Review by alex-ju — COMMENTED** [review]
> * Potential issues with the Dropdown `listPosition` and lack of collision awareness

While I see the issue here, I think this should be a concern of our Dropdown component, and something we can consider as a potential enhancement.

> * Button order and alignment at smaller viewport sizes

To me, having the same order makes sense regardless of their positioning. `flex-direction` it's certainly a no, as it will fail WCAG 1.3.2 Meaning sequence.

> * Responsive strategy
>   * Using container queries; are able to use these yet, does the support exist?

Kristin made a good point about checking in browser support on products. Although I suggest we see the usage of container queries in this component as a progressive enhancement – the layout doesn't break if the feature is not supported, it just looks slightly differently on narrow viewports.

>   * I've used `flex-wrap` with some of the inline elements, making the point at which they wrap variable. Is this something we want more control over?

Good question – curious to see others' views on this

>   * Is there a more elegant way to handle the breakpoints without using static values? Considering the method in the side-nav to conditionally apply classes with JavaScript, is that a recommended method here? Should the classes be exposed to be overwritten?

I _think_ that in the case of SideNav the applied classes were a way to instrument the component from outside, not necessarily a pattern that we need to replicate here.

>   * The linter does not recognize container queries yet, is that an adjustment that needs to be made?

Doesn't recognize them, but doesn't complaining either, so I say it's a non-issue.

> * I've written a few simple tests for this component, are more necessary? If so, I believe I can add them without too much trouble given that this is largely a presentational component.

Yes, I suggest we test the assertion (the error we throw when no title is provided), you'll be able to find some similar examples in other components.

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (alex-ju) [reply packages/components/addon/components/hds/page-header/index.hbs:11]: This was my recommendation, Melanie, to avoid nesting them, but if you think nesting the two landmarks is not a concern we can revert and apply the header element at the top level of this component.

**Comment by heatherlarsen** [comment]
> While I see the issue here, I think this should be a concern of our Dropdown component, and something we can consider as a potential enhancement.

@alex-ju I created [HDS-1995](https://hashicorp.atlassian.net/browse/HDS-1995) so we don't lose track of this work.

**Review by heatherlarsen — APPROVED** [review]
I know you didn't ask for a designer review on this, but I figured it could use some extra eyes since you're the original designer. This looks really solid and aligns with the Figma component well! :shipit:

**Comment by jorytindall** [comment]
> I know you didn't ask for a designer review on this, but I figured it could use some extra eyes since you're the original designer. This looks really solid and aligns with the Figma component well! :shipit:

Probably begs a question about who the approvers are in this process. I guess I'm my own stakeholder 😂

**Review by didoo — COMMENTED** [review]
Done a first pass and added a few comments (some of them may be blockers). I'll do a second review after the first round of reviews (I imagine there will be some tweaks to do)

Re. your request for feedback: 
- Potential issues with the Dropdown listPosition and lack of collision awareness
  - I agree with @alex-ju it's a separate issue, to be solved in another effort
- Using container queries
  - I agree with @alex-ju also on this: we should consider this as progressive enhancement, so I would not worry too much about it (in this very specific case, which is mostly presentational)
  - Is there a more elegant way to handle the breakpoints without using static values?
    - Not that I know/can think of
- Considering the method in the side-nav to conditionally apply classes with JavaScript, is that a recommended method here?
  - Nope :)
- The linter does not recognize container queries yet, is that an adjustment that needs to be made?
  -   Yes, but should be done in a separate ticket (it's about the tool we're using)

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (MelSumner) [reply packages/components/addon/components/hds/page-header/index.hbs:11]: It's acceptable to nest landmark elements directly 👍

File: packages/components/addon/components/hds/page-header/metadata.hbs
Comment (jorytindall) [reply packages/components/addon/components/hds/page-header/metadata.hbs:8]: Yes, this was initially intended to be a wrapper around the the metadata with other nested contextual components, but did not end up using that strategy. Good catch!

File: packages/components/app/styles/components/page-header.scss
Comment (jorytindall) [reply packages/components/app/styles/components/page-header.scss:81]: Ah, we had originally tried this to account for not rendering an empty `div` when no metadata is yielded, but alas it doesn't work. I'll remove it.

File: packages/components/app/styles/components/page-header.scss
Comment (jorytindall) [reply packages/components/app/styles/components/page-header.scss:64]: Ok, I've reworked these a bit to hopefully be a bit more semantic. Definitely open to suggestions if you have a recommendation, but there are less repetitive instances of "header/heading"

File: packages/components/addon/components/hds/page-header/index.hbs
Comment (jorytindall) [reply packages/components/addon/components/hds/page-header/index.hbs:20]: I was considering that, and upon further testing it _seems_ like it would make sense to use the same strategy as the rest of the yielded content and not pass the title as an argument. This is also the method used in the `Modal` and `Flyout`, and will make adoption easier. I'll make some adjustments and update the hierarchy.

File: packages/components/app/styles/components/page-header.scss
Comment (didoo) [inline packages/components/app/styles/components/page-header.scss:21]: I see there are two `container-type: inline-size` declaration (one here, and one in the parent). Are you sure you need this one here? If I comment it out in the devtools I don't see any effect 🤔

File: packages/components/app/styles/components/page-header.scss
Comment (didoo) [inline packages/components/app/styles/components/page-header.scss:20]: Niptick: some values are expressed as `rem`, others as `px`. Since the convention is to use `px` unless it's a font-size, can we convert all the `rem`s to `px`?

File: packages/components/tests/dummy/app/styles/app.scss
Comment (didoo) [inline packages/components/tests/dummy/app/styles/app.scss:1]: @jorytindall can you undo these changes? The quotes changes, I mean: in CSS we use double quotes for imports.

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/page-header.hbs:288]: could it be that it's now `768px`?

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/page-header.hbs:317]: Maybe instead of using a grid with three columns, what you could do is to have the same component inside three different containers, with different widths. Something like this:

```suggestion
      <Shw::Flex @direction="column" as |SF|>
        {{#let (array "399px" "767px" "100%") as |widths|}}
          {{#each widths as |width|}}
            <SF.Item @label="container width = {{width}}">
              <Shw::Outliner {{style width=width}}>
                <Hds::PageHeader as |PH|>
                  <PH.Title>Title of the page</PH.Title>
                  <PH.IconTile @icon="smartphone" />
                  <PH.Breadcrumb>
                    <Hds::Breadcrumb>
                      <Hds::Breadcrumb::Item @text="Organization" />
                      <Hds::Breadcrumb::Item @text="User" />
                    </Hds::Breadcrumb>
                  </PH.Breadcrumb>
                  <PH.Badges>
                    <Hds::Badge @text="Small viewport" @icon="tablet" />
                  </PH.Badges>
                  <PH.Subtitle>Forced small viewport</PH.Subtitle>
                  <PH.Actions>
                    <Hds::Button @text="Edit" @color="secondary" />
                    <Hds::Button @text="Create" @color="primary" />
                  </PH.Actions>
                </Hds::PageHeader>
              </Shw::Outliner>
            </SF.Item>
          {{/each}}
        {{/let}}
      </Shw::Flex>
```

**Review by didoo — COMMENTED** [review]
@jorytindall did a second pass on the code, added some more comments

File: packages/components/app/styles/components/page-header.scss
Comment (jorytindall) [reply packages/components/app/styles/components/page-header.scss:21]: Yes, looks like you're correct. Still figuring out the details of containers, but looks like declaring it once at the top of the component is sufficient.

File: packages/components/tests/dummy/app/styles/app.scss
Comment (jorytindall) [reply packages/components/tests/dummy/app/styles/app.scss:1]: Whoops, yes, sorry about that!

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (jorytindall) [reply packages/components/tests/dummy/app/templates/components/page-header.hbs:317]: That's much more elegant!

File: packages/components/app/styles/components/page-header.scss
Comment (didoo) [reply packages/components/app/styles/components/page-header.scss]: @jorytindall in case you didn't see it https://developer.mozilla.org/en-US/docs/Web/CSS/container-name (in case you need multiple "containers"; but in this case one is enough)

File: packages/components/app/styles/components/page-header.scss
Comment (didoo) [inline packages/components/app/styles/components/page-header.scss:52]: see comment above

```suggestion
.hds-page-header__title-badges-wrapper {
```

File: packages/components/tests/dummy/app/styles/app.scss
Comment (didoo) [inline packages/components/tests/dummy/app/styles/app.scss:26]: the two lines were on purpose :|

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/page-header.hbs:300]: We want to capture the wrapping when the page is snapshotted by Percy (to detect possible regressions in future) so the text needs to wrap at full-page width

```suggestion
        <PH.Title>Page title with a very very very very long string that can span multiple lines</PH.Title>
```

**Review by didoo — APPROVED** [review]
Left a few minor suggestions/comments, but apart from that it's good to go. Great job! 👍

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/page-header.hbs:215]:
```suggestion
        <PH.Description>Manage organization users, roles, and access control.</PH.Description>
        <PH.Description>And this is a second description, in case multiple paragraphs are needed.</PH.Description>
```

File: packages/components/tests/dummy/app/styles/app.scss
Comment (jorytindall) [reply packages/components/tests/dummy/app/styles/app.scss:26]: I believe this occurred at the same time as my changing the the double quotes to single quotes, which I think was prettier being a bit to heavy handed. I've been having issues with it especially related to markdown, let me take a look at my config.

File: packages/components/tests/dummy/app/templates/components/page-header.hbs
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/page-header.hbs:18]: Super nit: If content is indented it doesn't also need blank lines above and below the indented portion.

**Pull request merged**
